jax/api.py

```diff
@@ -1062,3 +1062,59 @@ def graphviz_maker(*args, **kwargs):
 
   graphviz_maker.__name__ = "make_graphviz({})".format(graphviz_maker.__name__)
   return graphviz_maker
+
+
+def eval_shape(fun, *args, **kwargs):
+  """Compute the shape of ``fun(*args, **kwargs)`` without incurring any FLOPs.
+
+  This utility function is useful for performing shape inference. Its
+  input/output behavior is defined by:
+
+    def eval_shape(fun, *args, **kwargs):
+      out = fun(*args, **kwargs)
+      return jax.tree_util.tree_map(np.shape, out)
+
+  But instead of applying ``fun`` directly, which might be expensive, it uses
+  JAX's abstract interpretation machinery to evaluate the shapes without doing
+  any FLOPs.
+
+  Using ``eval_shape`` can also catch shape errors, and will raise same shape
+  errors as evaluating ``fun(*args, **kwargs)``.
+
+  Args:
+    *args: a positional argument tuple of arrays, scalars, or (nested) standard
+      Python containers (tuples, lists, dicts, namedtuples, i.e. pytrees) of
+      those types. Since only the ``shape`` and ``dtype`` attributes are
+      accessed, only values that duck-type arrays are required, rather than real
+      ndarrays. The duck-typed objects cannot be namedtuples because those are
+      treated as standard Python containers. See the example below.
+    **kwargs: a keyword argument dict of arrays, scalars, or (nested) standard
+      Python containers (pytrees) of those types. As in ``args``, array values
+      need only be duck-typed to have ``shape`` and ``dtype`` attributes.
+
+  For example:
+
+  >>> f = lambda A, x: np.tanh(np.dot(A, x))
+  >>> class MyArgArray(object):
```

Comment:
While it's nice that it works with any object of this form, I feel like it'd be better if this library provided a class that does it -- just to help users.

Reply:
I hear you and agree 100% on your description of the tradeoff: API minimalism vs user convenience. My instinct is to err on the side of API minimalism for now, but we can revisit that as we see how this API gets used. It's easier to add something later than take it away. Also, while I know it's slightly more annoying to the user (i.e. you!) to have to make your own shape-and-dtype-struct class, in this case since it's only 4 lines and provided as a recipe in the docstring, the cost seems minimized.

In any case, hopefully this PR is a step in the right direction!